README
= Psensor README
:docinfo2:
:toc:
:icons: font

Introduction
------------

Psensor is a graphical hardware monitoring application for Linux.

It draws realtime charts and raises alerts about:

 * the temperature of the motherboard and CPU sensors (using
   lm-sensors).
 * the temperature of the NVidia GPUs (using XNVCtrl).
 * the temperature of ATI GPUs (using ATI ADL SDK).
 * the temperature of the Hard Disk Drives (using hddtemp, libatasmart
   or udisks2).
 * the rotation speed of the fans.
 * the temperature of a remote computer.
 * the CPU load.

Alerts are using Desktop Notification and a specific GTK+ status icon.

For Ubuntu users, Psensor is providing an Application Indicator which turns
red under alert and a menu for quickly check all sensors.

It is possible to monitor remotely a computer:

 * start +psensor-server+ on the remote computer.

 * start +psensor+ using: +psensor --url=http://hostname:3131+.

+psensor-server+ is a minimal Web server, a Web browser can be used to
monitor the remote computer by opening the URL:
+http://hostname:3131+.

[WARNING]
+psensor-server+ does not provide any way to restrict the connection
to the HTTP server, worst, no effort has been made against malicious
HTTP attacks.  You should make the +psensor-server+ port available
only to a network or computer you trust by using the usual network
security tools of the system (for example:
http://www.netfilter.org/projects/iptables/index.html[iptables]).

Installation Using Distribution Packages
----------------------------------------

Debian
~~~~~~

For Debian, the binary and source packages are available in the
Wheezy, unstable and testing repositories.

Debian package information is available at:
http://packages.qa.debian.org/p/psensor.html.

To do the installation:
--------------------------------------------
sudo apt-get install psensor psensor-server
--------------------------------------------

Ubuntu
~~~~~~

Since Ubuntu Oneiric (11.10), psensor is available in the universe
repository of Ubuntu.

To install +psensor+ or +psensor-server+:
------------------------------------
sudo apt-get install psensor
sudo apt-get install psensor-server
------------------------------------

Or search them in the Ubuntu Software Center.

Installation From Source Archive
--------------------------------

+psensor+ and +psensor-server+ can be compiled and installed on any
modern GNU/Linux distribution.

Compilation Prerequisites
~~~~~~~~~~~~~~~~~~~~~~~~~

The compilation of +psensor+ requires:

 * make
 * gcc (or llvm)
 * lm-sensors
 * library sensors4
 * library gtk3 >=3.4
 * help2man (optional, required to produce the manpage)
 * asciidoctor (optional, required to produce the html version of the
   faq)
 * cppcheck (optional, static source checker)
 * library libnotify (optional)
 * library libappindicator >= 3.0 (optional)
 * library libXNVCtrl (optional)
 * library json-c >= 0.11.99 and curl (optional, required for remote monitoring)
 * library unity (>=v3.4.2, optional)
 * library gtop2 (optional, required for CPU usage)
 * library atasmart (optional, for disk monitoring)
 * library udisk2 (optional, for disk monitoring)

The specific requirements for +psensor-server+:

 * library libmicrohttpd
 * library json-c >= 0.11.99

Compilation and Installation Steps
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. download the source archive from: http://wpitchoune.net/psensor/files
2. Extract files from the source archive
3. Compilation: +./configure;make clean all+
4. Installation:  +make install+ (require to have root permission)
5. Start the sensor detection script: +sensors-detect+ and follows the
instructions. Reboot or activate the kernel modules which have
been found.
6. Run: +psensor+

ATI/AMD GPU Support
~~~~~~~~~~~~~~~~~~~

When the OpenSource ATI driver is used, the monitoring
information is available through lm-sensors and does not require the
following instructions.

When the proprietary ATI driver(Catalyst) is used, the build is
requiring an additional library.

Download the
http://developer.amd.com/sdks/ADLSDK/Pages/default.aspx[ATI ADL SDK]
and extracts the files of the archive into a directory [dir].

Replace the step 3. of the above section by +./configure
--with-libatiadl=[dir]+ where +[dir]+ must be the directory where you
have extracted the ATI ADL SDK. Other steps are indentical.

Contact
-------

Bugs and comments can be sent to jeanfi@gmail.com.

Home page: http://wpitchoune.net/psensor


NEWS
= Psensor NEWS
:docinfo2:

v1.2.1
------

* Changed icons to better, higher resolution ones (Stuart Hayhurst)
* documentation files are generated with asciidoctor instead of
  asciidoc.
* added support of BCM2835 which is mostly used in the Raspberry PI3.
* Fresh install of 1.2.0 complains about missing sensor config
  keys. (LP: #1650378)
* Fixed hide on startup preference (Francis Chin). (LP: #1642029).
* Fixed restoration of the panel divider position (Francis Chin).
* Update graph immediately when a sensor is enabled or disabled (Mark S.).
* Moved the source code to GitLab: https://gitlab.com/jeanfi/psensor.git
* Fixed 'psensor-server crash when no sensor is detected' (Stéphane Gourichon).
* Fixed 'Compilation requires dependencies of all component even if only server is desired'.
* Fixed 'Any percent sensor breaks psensor-server page' (Stéphane Gourichon).
* Fixed potential race condition during graph drawing (Mark S.).

v1.2.0
------

* Added tooltip explaining why a setting is not available.
* Removed support of old AppIndicator library < 3.0.
* Fixed memory leaks in appindicator support when sensor preferences
  are changed.
* Changes of sensors preferences are applied immediately.
* Avoid multiple sensor preferences dialog at the same time.
* Support max/min sensor information from providers.
* Fixed string not translatable. (Closes: #138076).
* Added a default high threshold for temperature.
* Made help2man optional.
* New dconf settings +default-high-threshold-temperature+,
  and +default-sensor-alarm-enabled+.
* Fixed memory leaks due to glib GError.
* Added Greek translation.
* Updated Russian translation (Grigorii Chirkov).
* Updated German translation (Max).
* Added Polish translation (hawkeye116477).
* Fixed GtkDialog mapped without a transient parent (Francis
  Chin). (LP: #1642696).

v1.1.5
------

* Fixed CPU overuse. (LP: 1582930)

v1.1.4
------

* psensor-server: avoid to retrieve files which are not under the
  webserver directory.
* Fixed compilation failure due to MHD_create_response_from_buffer
  which has been deprecated

v1.1.3
------

* Fixed compilation issue with gtk+3.14 due to deprecation
  of GtkStatusIcon. (Closes: #763257).

v1.1.2
------

* Fixed missing description in psensor manpage.
* All mouse buttons can be used to raise color settings in the list of
  sensors.
* All mouse buttons can be used to popup menu in the list of sensors.
* List of sensors UI defined in glade file instead of code.
* Fixed graph enabled setting not saved when toggle in the list of
  sensors.
* Added NVidia GPU graphics, memory, PCIe and video usage.
* Fixed message when the configuration file is missing.
* Added NVidia fan support.
* Added 'Hide' menu item in the contextual menu of the list of
  sensors.
* Added free memory support.
* Added alarm support of sensors expressed as percent.
* Smoother curves (using Bezier algo).
* Added translator credits in the about dialog.
* Added Croatian translation from LP (gogo).
* Merged all translations from LP.
* Included the full list of contributors from LP, available throw the
  about dialog.
* Support of disk temperature using udisks2.
* Fixed small UI layout/padding issues.
* Splitted sensor preferences using a tab panel.
* New preferences to enable/disable each sensor provider.
* psensor: removed option +--use-libatasmart+.
* Changed default color for the curves (copied from the default
  colors of the gtk color chooser widget).

v1.1.1
------

* Sensor properties stored in +$HOME/.psensor/psensor.cfg+ instead of
  GConf.
* psensor settings stored with dconf, removed dependency to gconf2.
* Fixed check of sensors/error.h.
* Made asciidoc optional.
* Hide/show sensors in the main window.
* Added TODO in the source tarball.
* Merged Hungarian translation from LP (Richard Somlói).
* Merged Italian translation from LP (Gabriele).
* Merged Latvian translation from LP (tuxmaniack).
* Merged Swedish translation from LP (Josef Andersson).
* Meged Turkish translation from LP (İbrahim Çelik).

v1.0.4
------

* Fixed menu separators of the appindicator.

v1.0.3
------

* Fixed bug: psensor not autostarted when $HOME/.config/autostart does
  not exist.
* Fixed few swedish translations (Josef Andersson).
* Fixed psensor does not start: Attempt to unlock mutex that was not
  locked during startup with glib 2.41.2. (LP: #1346299).
* Fixed FTBFS with clang instead of gcc. (Closes: #753061).
* Fixed notification script call (wrong allocation of command string).

v1.0.2
------

* +psensor-server+:
  - Fixed file descriptor leaks.

v1.0.1
------

* Smaller and more consistent with ubuntu themes appindicator icons.
* Sensors displayed directly in the top bar (experimental).
* Improved status icon in hicolor theme to fit well with white
  background. (LP: #1310323).
* Psensor not autostarted by default.
* Added a setting to launch psensor on session startup. (Closes:
  #728754)
* Fixed minor GUI bugs (widget alignments, padding, layout policy,
  etc).
* Fixed missing translatable strings.
* Fixed FTBFS due to deprecation of json_object_object_get. (Closes:
  #746145)
* Fixed typo (Celcius/Celsius).
* Merged german translation from LP.
* Merged turkish translation from LP (İbrahim Çelik).
* Merged chinese translation from LP.
* Merged hungarian translation from LP.
* Merged russian translation from LP (Dmitry Serov).
* Added support of swedish translation (Josef Andersson).
* Fixed NVIDIA strings.
* Fixed default sensor color which was always black.
* Fixed compilation on ppc64. (Closes: #749110)

v0.8.0.6
--------

* Updated jquery and jqplot to last version.
* Removed excanvas for copyright reason (apache license not compatible
  with GPLv2 license).
* Use non-minimized version of .js and .css to avoid lintian errors
  (source-is-missing).

v0.8.0.5
--------

* Added NEWS.html.
* Converted NEWS file to asciidoc.
* +psensor-server+:
  - removed web fonts causing lintian warnings.
  - redesigned web pages.

src/ui_unity.c
/*
 * Copyright (C) 2010-2016 jeanfi@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */
#include <unity.h>

#include <cfg.h>
#include <temperature.h>
#include <ui_unity.h>

static UnityLauncherEntry * psensor_entry;
static bool count_visible;

static void
count_visible_changed_cbk(GSettings *settings, gchar *key, gpointer data)
{
	count_visible = config_is_count_visible();

	if (count_visible) {
		/* if the count is not set to non-0 value, it will not be visible */
		unity_launcher_entry_set_count(psensor_entry, 1);
		unity_launcher_entry_set_count_visible(psensor_entry, TRUE);
	} else {
		unity_launcher_entry_set_count_visible(psensor_entry, FALSE);
	}
}

static double get_max_current_value(struct psensor **sensors, unsigned int type)
{
	double m, v;
	struct psensor *s;

	m = UNKNOWN_DBL_VALUE;
	while (*sensors) {
		s = *sensors;

		if ((s->type & type) && config_is_sensor_graph_enabled(s->id)) {
			v = psensor_get_current_value(s);

			if (m == UNKNOWN_DBL_VALUE || v > m)
				m = v;
		}

		sensors++;
	}

	return m;
}

void ui_unity_launcher_entry_update(struct psensor **sensors)
{
	double v;

	if (!count_visible || !sensors || !*sensors)
		return;

	v = get_max_current_value(sensors, SENSOR_TYPE_TEMP);

	if (v != UNKNOWN_DBL_VALUE) {
		if (config_get_temperature_unit() == FAHRENHEIT)
			v = celsius_to_fahrenheit(v);

		unity_launcher_entry_set_count(psensor_entry, v);
	}
}

void ui_unity_init(void)
{
	psensor_entry = unity_launcher_entry_get_for_desktop_file
		(PSENSOR_DESKTOP_FILE);

	count_visible = config_is_count_visible();

	if (count_visible) {
		/* if the count is not set to non-0 value, it will not be visible */
		unity_launcher_entry_set_count(psensor_entry, 1);
		unity_launcher_entry_set_count_visible(psensor_entry, TRUE);
	} else {
		unity_launcher_entry_set_count_visible(psensor_entry, FALSE);
	}

	g_signal_connect_after(config_get_GSettings(),
			       "changed::interface-unity-launcher-count-disabled",
			       G_CALLBACK(count_visible_changed_cbk),
			       NULL);
}


src/ui.c
/*
 * Copyright (C) 2010-2018 jeanfi@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */
#include <stdlib.h>

#include <cfg.h>
#include <slog.h>
#include <ui.h>
#include <ui_appindicator.h>
#include <ui_graph.h>
#include <ui_pref.h>
#include <ui_sensorlist.h>
#include <ui_sensorpref.h>
#include <ui_status.h>

static GtkWidget * w_sensors_scrolled_tree;
static GtkWidget *w_graph;
static GtkContainer *w_sensor_box;
static GtkContainer *w_main_box;

static void update_layout(void)
{
	enum sensorlist_position sensorlist_pos;

	g_object_ref(w_sensors_scrolled_tree);
	g_object_ref(w_graph);

	gtk_container_remove(w_sensor_box,
			     w_sensors_scrolled_tree);

	gtk_container_remove(w_sensor_box, w_graph);

	gtk_container_remove(w_main_box, GTK_WIDGET(w_sensor_box));

	sensorlist_pos = config_get_sensorlist_position();
	if (sensorlist_pos == SENSORLIST_POSITION_RIGHT
	    || sensorlist_pos == SENSORLIST_POSITION_LEFT)
		w_sensor_box
			= GTK_CONTAINER(gtk_paned_new
					(GTK_ORIENTATION_HORIZONTAL));
	else
		w_sensor_box
			= GTK_CONTAINER(gtk_paned_new
					(GTK_ORIENTATION_VERTICAL));

	gtk_box_pack_end(GTK_BOX(w_main_box),
			 GTK_WIDGET(w_sensor_box), TRUE, TRUE, 2);

	if (sensorlist_pos == SENSORLIST_POSITION_RIGHT
	    || sensorlist_pos == SENSORLIST_POSITION_BOTTOM) {
		gtk_paned_pack1(GTK_PANED(w_sensor_box), w_graph, TRUE, TRUE);
		gtk_paned_pack2(GTK_PANED(w_sensor_box),
				w_sensors_scrolled_tree,
				FALSE,
				TRUE);
	} else {
		gtk_paned_pack1(GTK_PANED(w_sensor_box),
				w_sensors_scrolled_tree,
				FALSE,
				TRUE);
		gtk_paned_pack2(GTK_PANED(w_sensor_box), w_graph, TRUE, TRUE);
	}

	g_object_unref(w_sensors_scrolled_tree);
	g_object_unref(w_graph);

	gtk_widget_show_all(GTK_WIDGET(w_sensor_box));
}

static void set_decoration(GtkWindow *win)
{
	gtk_window_set_decorated(win, config_is_window_decoration_enabled());
}

static void set_keep_below(GtkWindow *win)
{
	gtk_window_set_keep_below(win, config_is_window_keep_below_enabled());
}

static void set_menu_bar_enabled(GtkWidget *bar)
{
	if (config_is_menu_bar_enabled())
		gtk_widget_show(bar);
	else
		gtk_widget_hide(bar);
}

static void
decoration_changed_cbk(GSettings *settings, gchar *key, gpointer data)
{
	set_decoration(GTK_WINDOW(data));
}

static void
keep_below_changed_cbk(GSettings *settings, gchar *key, gpointer data)
{
	set_keep_below(GTK_WINDOW(data));
}

static void
menu_bar_changed_cbk(GSettings *settings, gchar *key, gpointer data)
{
	set_menu_bar_enabled(GTK_WIDGET(data));
}

static void
sensorlist_position_changed_cbk(GSettings *settings, gchar *key, gpointer data)
{
	update_layout();
}

static void connect_cbks(GtkWindow *win, GtkWidget *menu_bar)
{
	log_fct_enter();

	g_signal_connect_after(config_get_GSettings(),
			       "changed::interface-window-decoration-disabled",
			       G_CALLBACK(decoration_changed_cbk),
			       win);

	g_signal_connect_after(config_get_GSettings(),
			       "changed::interface-window-keep-below-enabled",
			       G_CALLBACK(keep_below_changed_cbk),
			       win);

	g_signal_connect_after(config_get_GSettings(),
			       "changed::interface-menu-bar-disabled",
			       G_CALLBACK(menu_bar_changed_cbk),
			       menu_bar);

	g_signal_connect_after(config_get_GSettings(),
			       "changed::interface-sensorlist-position",
			       G_CALLBACK(sensorlist_position_changed_cbk),
			       menu_bar);


	log_fct_exit();
}

static void save_window_pos(struct ui_psensor *ui)
{
	gboolean visible;
	GtkWindow *win;
	struct config *cfg;

	visible = gtk_widget_get_visible(ui->main_window);
	log_debug("Window visible: %d", visible);

	if (visible == TRUE) {
		cfg = ui->config;

		win = GTK_WINDOW(ui->main_window);

		gtk_window_get_position(win, &cfg->window_x, &cfg->window_y);
		log_debug("Window position: %d %d",
			  cfg->window_x,
			  cfg->window_y);

		gtk_window_get_size(win,
				    &cfg->window_w,
				    &cfg->window_h);
		log_debug("Window size: %d %d", cfg->window_w, cfg->window_h);

		cfg->window_divider_pos
			= gtk_paned_get_position(GTK_PANED(w_sensor_box));

		config_save(cfg);
	}
}

static gboolean
on_delete_event_cb(GtkWidget *widget, GdkEvent *event, gpointer data)
{
	struct ui_psensor *ui = data;

	log_debug("is_status_supported: %d\n", is_status_supported());

	if (is_appindicator_supported() || is_status_supported()) {
		save_window_pos(ui);
		gtk_widget_hide(ui->main_window);
	} else {
		ui_psensor_quit(ui);
	}

	return TRUE;
}

void ui_show_about_dialog(GtkWindow *parent)
{
	static const char *const authors[] = { "jeanfi@gmail.com", NULL };

	log_fct("parent=%p", parent);

	gtk_show_about_dialog
		(parent,
		 "authors", authors,
		 "comments",
		 _("Psensor is a GTK+ application for monitoring hardware "
		   "sensors"),
		 "copyright",
		 _("Copyright(c) 2010-2020 jeanfi@gmail.com"),
#if GTK_CHECK_VERSION(3, 12, 0)
		 "license-type", GTK_LICENSE_GPL_2_0,
#endif
		 "logo-icon-name", "psensor",
		 "program-name", "Psensor",
		 "title", _("About Psensor"),
		 "translator-credits", _("translator-credits"),
		 "version", VERSION,
		 "website", PACKAGE_URL,
		 "website-label", _("Psensor Homepage"),
		 NULL);
}

void ui_cb_about(GtkAction *a, gpointer data)
{
	struct ui_psensor *ui;
	GtkWidget *parent;

	ui = (struct ui_psensor *)data;

	log_fct("ui=%p", ui);

	if (ui)
		parent = ui->main_window;
	else
		parent = NULL;

	ui_show_about_dialog(GTK_WINDOW(parent));
}

void ui_cb_menu_quit(GtkMenuItem *mi, gpointer data)
{
	ui_psensor_quit((struct ui_psensor *)data);
}

void ui_cb_preferences(GtkMenuItem *mi, gpointer data)
{
	ui_pref_dialog_run((struct ui_psensor *)data);
}

void ui_cb_sensor_preferences(GtkMenuItem *mi, gpointer data)
{
	struct ui_psensor *ui = data;

	if (ui->sensors && *ui->sensors)
		ui_sensorpref_dialog_run(*ui->sensors, ui);
}

void ui_psensor_quit(struct ui_psensor *ui)
{
	save_window_pos(ui);

	log_debug("Destroy main window");
	gtk_widget_destroy(ui->main_window);
	gtk_main_quit();
}

void ui_enable_alpha_channel(struct ui_psensor *ui)
{
	GdkScreen *screen;
	GdkVisual *visual;
	struct config *cfg;

	cfg = ui->config;

	screen = gtk_widget_get_screen(ui->main_window);

	log_debug("Config alpha channel enabled: %d",
		  cfg->alpha_channel_enabled);
	if (cfg->alpha_channel_enabled && gdk_screen_is_composited(screen)) {
		log_debug("Screen is composited");
		visual = gdk_screen_get_rgba_visual(screen);
		if (visual) {
			gtk_widget_set_visual(ui->main_window, visual);
		} else {
			cfg->alpha_channel_enabled = 0;
			log_err("Enable alpha channel has failed");
		}
	} else {
		cfg->alpha_channel_enabled = 0;
	}
}

static void slog_enabled_cbk(void *data)
{
	struct ui_psensor *ui;
	struct psensor **sensors;
	pthread_mutex_t *mutex;

	ui = (struct ui_psensor *)data;
	sensors = ui->sensors;
	mutex = &ui->sensors_mutex;

	log_debug("slog_enabled_cbk");

	if (is_slog_enabled())
		slog_activate(NULL, sensors, mutex, config_get_slog_interval());
	else
		slog_close();
}

void ui_window_create(struct ui_psensor *ui)
{
	GtkWidget *window, *menu_bar;
	GdkPixbuf *icon;
	GtkIconTheme *icon_theme;
	struct config *cfg;
	guint ok;
	GtkBuilder *builder;
	GError *error;

	log_fct("ui=%p", ui);

	builder = gtk_builder_new();

	error = NULL;
	ok = gtk_builder_add_from_file
		(builder,
		 PACKAGE_DATA_DIR G_DIR_SEPARATOR_S "psensor.glade",
		 &error);

	if (!ok) {
		log_printf(LOG_ERR, error->message);
		g_error_free(error);
		return;
	}

	window = GTK_WIDGET(gtk_builder_get_object(builder, "window"));
	gtk_builder_connect_signals(builder, ui);

	cfg = ui->config;

	config_set_slog_enabled_changed_cbk(slog_enabled_cbk, ui);

	gtk_window_set_default_size(GTK_WINDOW(window),
				    cfg->window_w,
				    cfg->window_h);

	icon_theme = gtk_icon_theme_get_default();
	icon = gtk_icon_theme_load_icon(icon_theme, "psensor", 48, 0, NULL);
	if (icon)
		gtk_window_set_icon(GTK_WINDOW(window), icon);
	else
		log_err(_("Failed to load Psensor icon."));

	g_signal_connect(window,
			 "delete_event", G_CALLBACK(on_delete_event_cb), ui);

	set_decoration(GTK_WINDOW(window));
	set_keep_below(GTK_WINDOW(window));

	menu_bar = GTK_WIDGET(gtk_builder_get_object(builder, "menu_bar"));
	w_main_box = GTK_CONTAINER(gtk_builder_get_object(builder, "main_box"));
	ui->popup_menu = GTK_WIDGET(gtk_builder_get_object(builder,
							   "popup_menu"));
	g_object_ref(G_OBJECT(ui->popup_menu));
	ui->main_window = window;
	w_graph = GTK_WIDGET(gtk_builder_get_object(builder, "graph"));
	ui_graph_create(ui);

	w_sensor_box = GTK_CONTAINER(gtk_builder_get_object(builder,
							    "sensor_box"));
	ui->sensors_store = GTK_LIST_STORE(gtk_builder_get_object
					   (builder, "sensors_store"));
	ui->sensors_tree = GTK_TREE_VIEW(gtk_builder_get_object
					 (builder, "sensors_tree"));
	w_sensors_scrolled_tree
		= GTK_WIDGET(gtk_builder_get_object
			     (builder, "sensors_scrolled_tree"));

	ui_sensorlist_create(ui);

	connect_cbks(GTK_WINDOW(window), menu_bar);

	log_debug("ui_window_create(): show_all");
	update_layout();
	gtk_widget_show_all(GTK_WIDGET(w_main_box));
	set_menu_bar_enabled(menu_bar);

	g_object_unref(G_OBJECT(builder));

	log_debug("ui_window_create() ends");
}

void ui_window_show(struct ui_psensor *ui)
{
	struct config *cfg;

	log_debug("ui_window_show()");

	cfg = ui->config;
	if (cfg->window_restore_enabled) {
		gtk_paned_set_position(GTK_PANED(w_sensor_box),
				       cfg->window_divider_pos);
		gtk_window_move(GTK_WINDOW(ui->main_window),
				cfg->window_x,
				cfg->window_y);
	}

	gtk_window_present(GTK_WINDOW(ui->main_window));
}

static int cmp_sensors(const void *p1, const void *p2)
{
	const struct psensor *s1, *s2;
	int pos1, pos2;

	s1 = *(void **)p1;
	s2 = *(void **)p2;

	pos1 = config_get_sensor_position(s1->id);
	pos2 = config_get_sensor_position(s2->id);

	return pos1 - pos2;
}

struct psensor **ui_get_sensors_ordered_by_position(struct psensor **sensors)
{
	struct psensor **result;

	result = psensor_list_copy(sensors);
	qsort(result,
	      psensor_list_size(result),
	      sizeof(struct psensor *),
	      cmp_sensors);

	return result;
}

GtkWidget *ui_get_graph(void)
{
	return w_graph;
}
